kobo/apps/audit_log/signals.py

```diff
@@ -24,3 +33,61 @@ def create_access_log(sender, user, **kwargs):
 def create_ph_log_for_import(sender, result, **kwargs):
     task = ImportTask.objects.get(uid=result)
     ProjectHistoryLog.create_from_import_task(task)
+
+
+def initialize_permission_lists_if_necessary(request):
+    if getattr(request, 'permissions_added', None) is None:
+        request.permissions_added = defaultdict(set)
+    if getattr(request, 'permissions_removed', None) is None:
+        request.permissions_removed = defaultdict(set)
+    if getattr(request, 'partial_permissions_added', None) is None:
+        request.partial_permissions_added = defaultdict(list)
+
+
+def initialize_request():
+    request = get_current_request()
+    if request is None:
+        return None
+    initialize_permission_lists_if_necessary(request)
+    return request
+
+
+@receiver(post_assign_perm, sender=Asset)
+def add_assigned_perms(sender, instance, user, codename, deny, **kwargs):
+    request = initialize_request()
+    if not request or instance.asset_type != ASSET_TYPE_SURVEY or deny:
+        return
+    request.permissions_added[user.username].add(codename)
+    request.permissions_removed[user.username].discard(codename)
+
+
+@receiver(post_remove_perm, sender=Asset)
+def add_removed_perms(sender, instance, user, codename, **kwargs):
+    request = initialize_request()
+    if not request or instance.asset_type != ASSET_TYPE_SURVEY:
+        return
+    request.permissions_removed[user.username].add(codename)
+    request.permissions_added[user.username].discard(codename)
+
+
+@receiver(post_assign_partial_perm, sender=Asset)
+def add_assigned_partial_perms(sender, instance, user, perms, **kwargs):
+    request = initialize_request()
+    if not request or instance.asset_type != ASSET_TYPE_SURVEY:
+        return
+    perms_as_list_of_dicts = [{'code': k, 'filters': v} for k, v in perms.items()]
+    # partial permissions are replaced rather than added
+    request.partial_permissions_added[user.username] = perms_as_list_of_dicts
+
+
+@receiver(post_remove_partial_perms, sender=Asset)
+def remove_partial_perms(sender, instance, user, **kwargs):
+    request = initialize_request()
+    if not request or instance.asset_type != ASSET_TYPE_SURVEY:
+        return
+    request.partial_permissions_added[user.username] = []
+    # in case we somehow deleted partial permissions
+    # without deleting 'partial_submissions', take care of that now since
+    # we can't have one without the other
+    request.permissions_added[user.username].discard(PERM_PARTIAL_SUBMISSIONS)
+    request.permissions_removed[user.username].add(PERM_PARTIAL_SUBMISSIONS)
```

Comment:
A -> Z 🙏

Reply:
I'm going to alphabetize everything in one big PR after all the functionality is done